### DISCUSSION TITLE
Matlab/JSBSim S-function: error while changing aircraft file

### DISCUSSION THREAD

**Discussion by MariaALJ** [JSBSim-Team/jsbsim#784] [Q&A]
Hello,

I am trying to change the aircraft in the JSBSim S-function under matlab but I always have the error : Aircraft file could not be loaded. For example, I'm going from the 737 (default) to the 787-8. I also tried to manually change the aircraft model in the configuration script that is called in the S-function besides the aircraft model, but it didn't help.

Any help on this matter would be really appreciated.

Thank you in adavance!

The error log:

JSBSim S-Function is initializing...

     JSBSim Flight Dynamics Model v1.1.12 [GitHub build 866/commit 133fcff53ae15abcbbbd4d06390e1b59ab76e5c6] Jun 26 2022 09:57:16
            [JSBSim-ML v2.0]
JSBSim startup beginning ...
Simulation dt set to 0.008333
	Setting up JSBSim with standard 'aircraft', 'engine', and 'system' paths.
	Loading aircraft '787-8' ...
Reading Aircraft Configuration File: 787-8
                            Version: 2.0
This aircraft model is a BETA release!!!
This aircraft model probably will not fly as expected.
Use this model for development purposes ONLY!!!
  Description:   Flight Dynamics Model for the Boeing 787-8
  Model Author:  Narendran Muraleedharan, Dwayne Gable and Nicky Ivy
  Creation Date: 2011-12-29
  Version:       $Revision: 1.10 $
  Aircraft Metrics:
    WingArea: 3501.798400
    WingSpan: 192.000000
    Incidence: 0.034907
    Chord: 38.940000
    H. Tail Area: 343.360000
    H. Tail Arm: 81.600000
    V. Tail Area: 471.777280
    V. Tail Arm: 0.000000
    Eyepoint (x, y, z): 0.000000 , 0.000000 , 0.000000
    Ref Pt (x, y, z): 0.000000 , 0.000000 , 0.000000
    Visual Ref Pt (x, y, z): 0.000000 , 0.000000 , 0.000000
  Mass and Balance:
    baseIxx: 7675399.200088 slug-ft2
    baseIyy: 6595543.304209 slug-ft2
    baseIzz: 22197429.613715 slug-ft2
    baseIxy: -0.015489 slug-ft2
    baseIxz: 0.015489 slug-ft2
    baseIyz: -0.000000 slug-ft2
    Empty Weight: 239200.000000 lbm
    CG (x, y, z): -16.000000 , 0.000000 , 22.597200
    Point Mass Object: 350.000000 lbs. at X, Y, Z (in.): -898.320000  0.000000  51.220000
    Point Mass Object: 1800.000000 lbs. at X, Y, Z (in.): -591.000000  0.000000  78.800000
    Point Mass Object: 6400.000000 lbs. at X, Y, Z (in.): -216.700000  0.000000  78.800000
    Point Mass Object: 25000.000000 lbs. at X, Y, Z (in.): 157.600000  0.000000  78.800000
    Point Mass Object: 10000.000000 lbs. at X, Y, Z (in.): -394.000000  0.000000  -39.400000
    Point Mass Object: 5000.000000 lbs. at X, Y, Z (in.): 157.600000  0.000000  -39.400000
  Ground Reactions:
    BOGEY NOSE_GEAR
      Location: -809.500580 , 0.000000 , -143.368720
      Spring Constant:  160000.000000
      Damping Constant: 80000.000000 (linear)
      Rebound Damping Constant: 80000.000000 (linear)
      Dynamic Friction: 0.300000
      Static Friction:  0.400000
      Rolling Friction: 0.020000
      Steering Type:    STEERABLE
      Grouping:         NONE
      Max Steer Angle:  70.000000
      Retractable:      1
    BOGEY LEFT_MAIN
      Location: 77.105800 , -193.599780 , -153.632420
      Spring Constant:  833000.000000
      Damping Constant: 166600.000000 (linear)
      Rebound Damping Constant: 166600.000000 (linear)
      Dynamic Friction: 0.350000
      Static Friction:  0.400000
      Rolling Friction: 0.020000
      Steering Type:    FIXED
      Grouping:         LEFT
      Max Steer Angle:  0.000000
      Retractable:      1
    BOGEY RIGHT_MAIN
      Location: 77.105800 , 193.599780 , -153.632420
      Spring Constant:  833000.000000
      Damping Constant: 166600.000000 (linear)
      Rebound Damping Constant: 166600.000000 (linear)
      Dynamic Friction: 0.350000
      Static Friction:  0.400000
      Rolling Friction: 0.020000
      Steering Type:    FIXED
      Grouping:         RIGHT
      Max Steer Angle:  0.000000
      Retractable:      1
    STRUCTURE LEFT_WING_TIP
      Location: 549.236000 , -1173.481720 , 154.830180
      Spring Constant:  100000.000000
      Damping Constant: 20000.000000 (linear)
      Rebound Damping Constant: 20000.000000 (linear)
      Dynamic Friction: 0.400000
      Static Friction:  0.500000
    STRUCTURE RIGHT_WING_TIP
      Location: 549.236000 , 1173.481720 , 154.830180
      Spring Constant:  100000.000000
      Damping Constant: 20000.000000 (linear)
      Rebound Damping Constant: 20000.000000 (linear)
      Dynamic Friction: 0.400000
      Static Friction:  0.500000
    STRUCTURE TAIL_STRIKE
      Location: 924.938640 , 0.000000 , 3.419920
      Spring Constant:  100000.000000
      Damping Constant: 20000.000000 (linear)
      Rebound Damping Constant: 20000.000000 (linear)
      Dynamic Friction: 0.400000
      Static Friction:  0.500000
    STRUCTURE NOSE
      Location: -984.290800 , 0.000000 , -29.845500
      Spring Constant:  100000.000000
      Damping Constant: 20000.000000 (linear)
      Rebound Damping Constant: 20000.000000 (linear)
      Dynamic Friction: 0.400000
      Static Friction:  0.500000
  External Reactions:
    Frame: BODY
    Location: (0.000000, 0.000000, 0.000000)
  Propulsion:
      FUEL tank holds 52000.000000 lbs. FUEL
      currently at 63.594883% of maximum capacity
      Tank location (X, Y, Z): 0.000000, -295.500000, 25.610000
      Effective radius: 0.000000 inches
      Initial temperature: -9999.000000 Fahrenheit
      Priority: 1
      FUEL tank holds 52000.000000 lbs. FUEL
      currently at 63.594883% of maximum capacity
      Tank location (X, Y, Z): 0.000000, 295.500000, 25.610000
      Effective radius: 0.000000 inches
      Initial temperature: -9999.000000 Fahrenheit
      Priority: 1
      FUEL tank holds 120000.000000 lbs. FUEL
      currently at 55.115566% of maximum capacity
      Tank location (X, Y, Z): 0.000000, 0.000000, -39.400000
      Effective radius: 0.000000 inches
      Initial temperature: -9999.000000 Fahrenheit
      Priority: 1
Simulation completed.
Remember to reset the program by typing clearSF in the matlab command window!
Error reported by S-function 'JSBSim_SFunction' in '[ex737cruise/S-Function](matlab:open_and_hilite_hyperlink ('ex737cruise/S-Function','error'))':
Aircraft file could not be loaded.

**Comment by bcoconni** [comment]
That's weird indeed. I can load the 787-7 model successfully using JSBSim Python interface (I don't have MATLAB installed on my computer). The error seems to occur while loading the tables `IdleThrust` in the file `engine/trent_1000.xml`: I'm getting some output from JSBSim for these tables that are now listed in your log.
```
      Tank location (X, Y, Z): 0, 0, -39.4
      Effective radius: 0 inches
      Initial temperature: -9999 Fahrenheit
      Priority: 1
    2 dimensional table with 6 rows, 8 columns.
                -10000.0000     0.0000  10000.0000      20000.0000      30000.0000      40000.0000      50000.0000      60000.0000
        0.0000  0.0430  0.0488  0.0528  0.0694  0.0899  0.1183  0.1467  0.0000
        0.2000  0.0500  0.0501  0.0335  0.0544  0.0797  0.1049  0.1342  0.0000
        0.4000  0.0040  0.0047  0.0020  0.0272  0.0595  0.0891  0.1203  0.0000
        0.6000  0.0000  0.0000  0.0000  0.0000  0.0276  0.0718  0.1073  0.0000
        0.8000  0.0000  0.0000  0.0000  0.0000  0.0474  0.0868  0.0900  0.0000
        1.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0552  0.0800  0.0000
    Function: propulsion/engine[0]/IdleThrust
    2 dimensional table with 8 rows, 8 columns.
                -10000.0000     0.0000  10000.0000      20000.0000      30000.0000      40000.0000      50000.0000      60000.0000
        0.0000  1.2600  1.0000  0.7400  0.5340  0.3720  0.2410  0.1490  0.0000
        0.2000  1.1710  0.9340  0.6970  0.5060  0.3550  0.2310  0.1430  0.0000
        0.4000  1.1500  0.9210  0.6920  0.5060  0.3570  0.2330  0.1450  0.0000
        0.6000  1.1810  0.9510  0.7210  0.5320  0.3780  0.2480  0.1540  0.0000
        0.8000  1.2580  0.9000  0.7020  0.5220  0.3770  0.2750  0.1700  0.0000
        1.0000  1.3690  1.1000  0.8510  0.6410  0.4550  0.3050  0.1960  0.0000
        1.2000  1.4850  1.2300  0.9750  0.7440  0.5450  0.3640  0.2250  0.0000
        1.4000  1.5941  1.3400  1.0860  0.8450  0.6280  0.4240  0.2630  0.0000
    Function: propulsion/engine[0]/MilThrust

In file ../../../aircraft/787-8/787-8.xml: line 251
Engine location ignored, only thruster location is used.

In file ../../../aircraft/787-8/787-8.xml: line 256
Engine orientation ignored, only thruster orientation is used.
      X = -211.9208
      Y = -383.1571
      Z = -31.2915
      Pitch = 0.0000 degrees
      Yaw = 0.0000 degrees
```

@noah-curran, any ideas ?

**Reply by noah-curran** [reply]
Off the top of my head, no ideas. I've used the S-Function for multiple custom aircraft, so I'm certain it is capable of using other aircraft besides the default 737. Without more information, I'm inclined to believe there is an issue with the aircraft file itself:
```
Loading aircraft '787-8' ...
Reading Aircraft Configuration File: 787-8
Version: 2.0
This aircraft model is a BETA release!!!
This aircraft model probably will not fly as expected.
Use this model for development purposes ONLY!!!
```
These warning messages seem awfully ominous...

**Reply by bcoconni** [reply]
> ```
> Loading aircraft '787-8' ...
> Reading Aircraft Configuration File: 787-8
> Version: 2.0
> This aircraft model is a BETA release!!!
> This aircraft model probably will not fly as expected.
> Use this model for development purposes ONLY!!!
> ```
> 
> These warning messages seem awfully ominous...

Nothing sinister there: that's the standard disclaimer for WIP flight models.

Anyway as I mentioned previously I can load this model successfully using Python so it's unlikely the model itself is at fault. I'd rather bet for a setting in the S-Function that needs some adjustment ?

**Reply by noah-curran** [reply]
I think I'm confused. You said you loaded in the 787-7 and @MariaALJ said she used the 787-8. Are these two models not different?

**Reply by noah-curran** [reply]
Regardless, I'm also still confused why you believe the S-Function is at fault here. The S-Function does not meddle with the aircraft files loading, it just calls JSBSim to load it. If an error was thrown, then the S-Function captured that the aircraft file was not properly loaded by JSBSim and terminated execution.
[See these lines of code:](https://github.com/JSBSim-Team/jsbsim/blob/master/matlab/JSBSim_SFunction.cpp#L437-L440)
```C
        if (!JII->OpenAircraft(aircraft)){
            ssSetErrorStatus(S, "Aircraft file could not be loaded.\n");
            return;
        }
```
I can dive further later when I take my closer look at overhauling parts of the S-Function's error returns, but I don't think there is anything definitive here that the S-Function is at fault. If the problem is that the error messages aren't descriptive enough, I agree completely re the discussion in #774. I'd advise making sure the aircraft file is OK.

**Reply by bcoconni** [reply]
> I think I'm confused. You said you loaded in the 787-7 and @MariaALJ said she used the 787-8. Are these two models not different?

Oops ! Sorry that was a typo, I meant that I have tested the "787-8" model.
> I can dive further later when I take my closer look at overhauling parts of the S-Function's error returns, but I don't think there is anything definitive here that the S-Function is at fault.

I did not mean that the S-Function code was at fault but rather that may be one of the S-Function input needs to be modified in addition to changing the model name ? I was thinking to initial conditions for instance ?
